Extensible mechanism to provide different serialization method

thepirat000 opened the pull request:
An initial implementation for an Extensible mechanism to provide a different serialization method.

To use custom serialization methods, add an Application Setting with key _RedisSerializerType_ containing the type name of a class that implements the interface **Microsoft.Web.Redis.ISerializer**.

For example:

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <appSettings>
    <add key="RedisSerializerType" value="MyCompany.Redis.JsonSerializer"/>
  </appSettings>
</configuration>

```

An implementation for a JSon serializer using [JSON.NET](http://www.newtonsoft.com/json) could be:
```
namespace MyCompany.Redis
{
    public class JsonSerializer : ISerializer
    {
        private static JsonSerializerSettings _settings = new JsonSerializerSettings() { TypeNameHandling = TypeNameHandling.All };

        public byte[] Serialize(object data)
        {
            return Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(data, _settings));
        }

        public object Deserialize(byte[] data)
        {
            if (data == null)
            {
                return null;
            }
            return JsonConvert.DeserializeObject(Encoding.UTF8.GetString(data), _settings);
        }
    }
}
```

azurecla commented:
Hi __@thepirat000__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.azure.com.
    </span>

TTYL, AZPRBOT;

SiddharthChatrolaMs commented:
Please also add test cases. Please sign a contribution license agreement.

SiddharthChatrolaMs commented on src/Shared/RedisUtility.cs:
```diff
@@ -5,13 +5,24 @@
+
+        static RedisUtility()
+        {
+            var serializerType = ConfigurationManager.AppSettings["RedisSerializerType"];
```
Make "RedisSerializerType" a redis session state and output cache parameter. so it will follow same path as other parameters.

thepirat000 replied:
I think that to make that change viable we'll to convert the RedisUtility static class to a non-static class in order to inject the provider configuration.
Do you agree?

thepirat000 commented:
Just changed the configuration key to be a redis session state and output cache parameter.

For example:

```
<sessionState mode="Custom" customProvider="MySessionStateStore">
    <providers>
        <add type = "Microsoft.Web.Redis.RedisSessionStateProvider"
             name = "MySessionStateStore"
             redisSerializerType = "MyCompany.Redis.JsonSerializer,MyCompanyDll"
             ... />
    </providers>
</sessionState>
```

Also added some unit tests and edited the [configuration page](https://github.com/Azure/aspnet-redis-providers/wiki/Configuration) on wiki.

SiddharthChatrolaMs commented:
Please sign the "CLA" at "https://cla.azure.com.". I can't merge before that.

thepirat000 commented:
I already signed it before, but anyway I just signed it again

![image](https://cloud.githubusercontent.com/assets/9836380/17455225/d9e76d80-5b76-11e6-8d3c-4f0fb79f0d64.png)

azurecla commented:
Hi __@thepirat000__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

SiddharthChatrolaMs commented:
NuGet packages are out: 
https://www.nuget.org/packages/Microsoft.Web.RedisSessionStateProvider/2.2.2
https://www.nuget.org/packages/Microsoft.Web.RedisOutputCacheProvider/1.7.2
